Beam/Include/Beam/Queries/AsyncDataStore.hpp

```diff
@@ -0,0 +1,295 @@
+#ifndef BEAM_ASYNC_DATA_STORE_HPP
+#define BEAM_ASYNC_DATA_STORE_HPP
+#include <memory>
+#include <vector>
+#include <boost/noncopyable.hpp>
+#include <boost/thread/locks.hpp>
+#include <boost/thread/mutex.hpp>
+#include "Beam/IO/OpenState.hpp"
+#include "Beam/Pointers/Dereference.hpp"
+#include "Beam/Pointers/LocalPtr.hpp"
+#include "Beam/Queries/Queries.hpp"
+#include "Beam/Queries/LocalDataStore.hpp"
+#include "Beam/Queues/RoutineTaskQueue.hpp"
+#include "Beam/Utilities/Algorithm.hpp"
+
+namespace Beam::Queries {
+namespace Details {
+  template<typename T>
+  class BaseMatches {
+    public:
+      virtual T AcquireNext() = 0;
+      virtual const T& PeekNext() const = 0;
+      virtual bool IsDepleted() const = 0;
+      virtual ~BaseMatches() {};
+  };
+
+  template<typename I>
+  class Matches final : public BaseMatches<typename I::value_type> {
+    public:
+      using Iterator = I;
+      using Type = typename Iterator::value_type;
+
+      Matches(Iterator begin, Iterator end)
+        : m_begin(std::move(begin)),
+          m_end(std::move(end)),
+          m_cur(m_begin + 1),
+          m_nextValue(*m_begin) {}
+
+      Type AcquireNext() override final {
+        auto match = std::move(m_nextValue);
+        if(m_cur == m_end) {
+          m_cur = m_begin;
+        } else {
+          m_nextValue = *m_cur;
+          ++m_cur;
+        }
+        return match;
+      }
+
+      const Type& PeekNext() const override final {
+        return m_nextValue;
+      }
+
+      bool IsDepleted() const override final {
+        return m_cur == m_begin;
+      }
+
+    private:
+      Iterator m_begin;
+      Iterator m_end;
+      Iterator m_cur;
+      Type& m_nextValue;
+  };
+
+  template<typename I>
+  std::unique_ptr<Matches<I>> MakeMatches(I begin, I end) {
+    return std::make_unique<Matches<I>>(std::move(begin), std::move(end));
+  }
+
+  template<typename T>
+  std::tuple_element_t<0, T> MergeMatches(T matches,
+      const SnapshotLimit& limit) {
+    using Type = typename std::tuple_element_t<0, T>::value_type;
+    auto data = std::list<std::unique_ptr<BaseMatches<Type>>>();
+    std::apply([&] (auto&... parts) {
+      auto testInclude = [&] (auto& part) {
+        if(!part.empty()) {
+          if(limit.GetType() == SnapshotLimit::Type::HEAD) {
+            data.push_back(MakeMatches(part.begin(), part.end()));
+          } else {
+            data.push_back(MakeMatches(part.rbegin(), part.rend()));
+          }
+        }
+      };
+      (testInclude(parts), ...);
+    }, std::move(matches));
+    auto result = std::vector<Type>();
+    for(auto i = 0; !data.empty() && i < limit.GetSize(); ++i) {
+      auto partIter = data.begin();
+      auto comparator = [](auto& lhs, auto& rhs) {
+        return SequenceComparator()(lhs->PeekNext(), rhs->PeekNext());
+      };
+      if(limit.GetType() == SnapshotLimit::Type::HEAD) {
+        partIter = std::min_element(data.begin(), data.end(), comparator);
```

Comment:
Is this neccessary? One property of DataStore's is that they always return values sorted in ascending order by sequence number, so this may not be needed.